src/platform/Linux/MdnsImpl.cpp

```diff
@@ -0,0 +1,665 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "MdnsImpl.h"
+
+#include <algorithm>
+#include <sstream>
+#include <string.h>
+#include <time.h>
+#include <vector>
+
+#include <netinet/in.h>
+
+#include "platform/Mdns.h"
+#include "support/CodeUtils.h"
+
+using std::chrono::duration_cast;
+using std::chrono::microseconds;
+using std::chrono::seconds;
+using std::chrono::steady_clock;
+
+namespace chip {
+namespace DeviceLayer {
+
+MdnsAvahi MdnsAvahi::sInstance;
+
+constexpr uint64_t kUsPerSec = 1000 * 1000;
+
+Poller::Poller(void)
+{
+    mAvahiPoller.userdata         = this;
+    mAvahiPoller.watch_new        = WatchNew;
+    mAvahiPoller.watch_update     = WatchUpdate;
+    mAvahiPoller.watch_get_events = WatchGetEvents;
+    mAvahiPoller.watch_free       = WatchFree;
+
+    mAvahiPoller.timeout_new    = TimeoutNew;
+    mAvahiPoller.timeout_update = TimeoutUpdate;
+    mAvahiPoller.timeout_free   = TimeoutFree;
+}
+
+AvahiWatch * Poller::WatchNew(const struct AvahiPoll * poller, int fd, AvahiWatchEvent event, AvahiWatchCallback callback,
```

Comment:
Move these static wrapper functions to an unnamed namespace & remove them from the class.

Reply:
The poller has to be in the header file since the PlatformManager has to access the file descriptors inside the poll to drive the event loop.

Reply:
PlatformManager just calls 

  `chip::Protocols::Mdns::ProcessMdns(mReadSet, mWriteSet, mErrorSet);`

I don't see Poller used anywhere outside this file.

What am I missing?

Reply:
We have to find a place to locate the data structure. Currently it's located in the `MdnsAvahi` class so we need its definition in the header file. Moving it to the source file means another static variable or allocating from the heap which I'm not willing to introduce.